File: trunk/Extensions/Widgets/ResizeHandler.cs
/*
 * Ra-Ajax - A Managed Ajax Library for ASP.NET and Mono
 * Copyright 2008 - Thomas Hansen thomas@ra-ajax.org
 * This code is licensed under the LGPL version 3 which 
 * can be found in the license.txt file on disc.
 * 
 */

using System;
using System.ComponentModel;
using WEBCTRLS = System.Web.UI.WebControls;
using ASP = System.Web.UI;
using Ra.Widgets;
using System.IO;
using HTML = System.Web.UI.HtmlControls;

[assembly: ASP.WebResource("Extensions.Js.ResizeHandler.js", "text/javascript")]

namespace Ra.Extensions
{
    /**
     * Ajax timer, raises Tick evnt handler back to server periodically. Alternative to Comet component and
     * far "safer" to use than Comet
     */
    [ASP.ToolboxData("<{0}:ResizeHandler runat=\"server\" />")]
    public class ResizeHandler : RaControl, IRaControl
    {
        /**
         * EventArgs passed when Viewport is resized
         */
        public class ResizedEventArgs : EventArgs
        {
            private int _width;
            private int _height;

            internal ResizedEventArgs(int width, int height)
            {
                _width = width;
                _height = height;
            }

            public int Width
            {
                get { return _width; }
            }

            public int Height
            {
                get { return _height; }
            }
        }

        /**
         * Raised periodically every Duration milliseconds
         */
        public event EventHandler<ResizedEventArgs> Resized;

        protected override void OnInit(EventArgs e)
        {
            base.OnInit(e);
            AjaxManager.Instance.IncludeScriptFromResource(typeof(Timer), "Extensions.Js.ResizeHandler.js");
        }

        void IRaControl.DispatchEvent(string name)
        {
            switch (name)
            {
                case "resized":
                    if (Resized != null)
                        Resized(this, 
                            new ResizedEventArgs(
                                Int32.Parse(Page.Request.Params["width"]), 
                                Int32.Parse(Page.Request.Params["height"])));
                    break;
            }
        }

		protected override string GetClientSideScriptType()
		{
			return "new Ra.ResHand";
		}

        protected override string GetOpeningHTML()
        {
            // Dummy HTML DOM element to make registration and such easier...
            return string.Format("<input type=\"hidden\" id=\"{0}\" />", ClientID);
        }
    }
}

File: Ra.Extensions/Js/ResizeHandler.js
/*
 * Ra-Ajax - A Managed Ajax Library for ASP.NET and Mono
 * Copyright 2008 - 2009 - Thomas Hansen thomas@ra-ajax.org
 * This code is licensed under the GPL version 3 which 
 * can be found in the license.txt file on disc.
 * 
 */


(function(){

// Creating class
Ra.ResHand = Ra.klass();


// Inheriting from Ra.Control
Ra.extend(Ra.ResHand.prototype, Ra.Control.prototype);


// Creating IMPLEMENTATION of class
Ra.extend(Ra.ResHand.prototype, {
  init: function(el, opt) {
    this.initControl(el, opt);
    Ra.extend(window, Ra.Element.prototype);
    window.observe('resize', this.onResized, this);
    this._size = {      width: window.innerWidth || document.body.clientWidth,       height: window.innerHeight || document.body.clientHeight    };    this.callback(this._size);
  },
  onResized: function() {    this._size = {      width: window.innerWidth || document.body.clientWidth,       height: window.innerHeight || document.body.clientHeight    };    var T = this;    setTimeout(function(){      T.checkToSeeIfCallback();    }, 500);  },
  checkToSeeIfCallback: function() {
    var sz = {      width: window.innerWidth || document.body.clientWidth,       height: window.innerHeight || document.body.clientHeight    };    if( sz.width == this._size.width && sz.height == this._size.height ) {      this.callback(this._size);    } else {      this._size = sz;      var T = this;      setTimeout(function(){        T.checkToSeeIfCallback();      }, 500);    }  },

  callback: function(sz) {
    new Ra.Ajax({
      args:'__RA_CONTROL=' + this.element.id + '&__EVENT_NAME=resized' + '&width=' + sz.width + '&height=' + sz.height,
      raCallback:true,
      onSuccess: this.onFinishedRequest,
      callingContext: this
    });
  }});})();
